src/librustc_typeck/collect.rs

```diff
@@ -1682,28 +1682,65 @@ fn find_existential_constraints(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
     }
 }
 
+pub fn get_infer_ret_ty(output: &'_ hir::FunctionRetTy) -> Option<&hir::Ty> {
+    if let hir::FunctionRetTy::Return(ref ty) = output {
+        if let hir::TyKind::Infer = ty.node {
+            return Some(&**ty)
+        }
+    }
+    None
+}
+
 fn fn_sig(tcx: TyCtxt<'_>, def_id: DefId) -> ty::PolyFnSig<'_> {
     use rustc::hir::*;
     use rustc::hir::Node::*;
 
     let hir_id = tcx.hir().as_local_hir_id(def_id).unwrap();
 
-    let icx = ItemCtxt::new(tcx, def_id);
```

Comment:
I don't see any need for moving this in.

Reply:
Yeah, initially merged the two branches into one and therefore moved the value in, but ended up splitting them anyways, therefore, no need for this change.